src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -11849,24 +11849,37 @@ ExpressionSyntax parseWhenNotNull(ExpressionSyntax expr)
             {
                 while (true)
                 {
-                    // Nullable suppression operators should only be consumed by a conditional access
-                    // if there are further conditional operations performed after the suppression
-                    if (isOptionalExclamationsFollowedByConditionalOperation())
-                    {
-                        while (this.CurrentToken.Kind == SyntaxKind.ExclamationToken)
-                            expr = _syntaxFactory.PostfixUnaryExpression(SyntaxKind.SuppressNullableWarningExpression, expr, EatToken());
-                    }
+                    // We should consume suppression '!'s which are in the middle of the 'whenNotNull', but not at the end.
+                    // For example, 'a?.b!.c' should be a cond-access whose RHS is '.b!.c',
+                    // while 'a?.b!' should be a suppression-expr containing a cond-access 'a?.b'.
+                    using var beforeSuppressionsResetPoint = GetDisposableResetPoint(resetOnDispose: false);
+
+                    while (this.CurrentToken.Kind == SyntaxKind.ExclamationToken)
+                        expr = _syntaxFactory.PostfixUnaryExpression(SyntaxKind.SuppressNullableWarningExpression, expr, EatToken());
 
-                    // Expand to consume the `dependent_access *` continuations.
+                    // Expand to consume the `dependent_access*` continuations.
                     if (tryParseDependentAccess(expr) is ExpressionSyntax expandedExpression)
                     {
                         expr = expandedExpression;
                         continue;
                     }
 
+                    // a?.b?.c
+                    // a?.b!?.c
                     if (TryParseConditionalAccessExpression(expr, out var conditionalAccess))
                         return conditionalAccess;
 
+                    // a?.b = c
+                    // a?.b! = c
+                    var (operatorTokenKind, operatorExpressionKind) = GetExpressionOperatorTokenKindAndExpressionKind();
+                    if (IsExpectedAssignmentOperator(operatorTokenKind))
+                    {
+                        return ParseAssignmentExpression(operatorExpressionKind, expr, EatExpressionOperatorToken(operatorTokenKind));
+                    }
```

Comment:
just to check, this is correc there, and shouldn't be in tryParseDependentAccess?

Reply:
Yeah this is similar to the trailing cond-access case, in that you will never parse out an assignment operator, then go on to parse out a further dependent-access. The [feature grammar](https://github.com/dotnet/csharplang/blob/main/proposals/null-conditional-assignment.md#specification) is meant to imply this but could possibly be made more clear.

Reply:
sgtm.  might be worth a comment to indicate that this ensures that we parse in the following fashion: 

```c#
return a?.b = c?.d = e?.f; // equivalent to:
return a?.b = (c?.d = e?.f); // equivalent to:
```